packages/wallet/src/TxBuilder/OutputBuilder.ts

```diff
@@ -0,0 +1,106 @@
+import { Cardano } from '@cardano-sdk/core';
+
+import {
+  DeepPartial,
+  MaybeValidTxOut,
+  OutputBuilder,
+  OutputValidationMinimumCoin,
+  OutputValidationMissingRequired,
+  OutputValidationTokenBundleSize
+} from './types';
+import { OutputValidation } from '../types';
+import { OutputValidator } from '../services';
+
+// Typescript is complaining about DeepPartial<Cardano.TokenMap> methods being optional
+// Narrow down DeepPartial<Cardano.TokenMap> to Cardano.TokenMap
+const isTokenMap = (map: DeepPartial<Cardano.TokenMap> | undefined): map is Cardano.TokenMap => !!map?.set;
+const isViableTxOut = (txOut: DeepPartial<Cardano.TxOut>): txOut is Cardano.TxOut =>
+  !!(txOut?.address && txOut?.value?.coins); // txOut must have at least an address and coins
+
+export const toOutputValidationError = (
+  txOut: Cardano.TxOut,
+  validation: OutputValidation
+): OutputValidationMinimumCoin | OutputValidationTokenBundleSize | undefined => {
+  if (validation.coinMissing) {
+    return new OutputValidationMinimumCoin(txOut, validation);
+  }
+  if (validation.tokenBundleSizeExceedsLimit) {
+    return new OutputValidationTokenBundleSize(txOut, validation);
+  }
+};
+
+export class OutputBuilderSaw implements OutputBuilder {
+  partialOutput: DeepPartial<Cardano.TxOut>;
+
+  #outputValidator: OutputValidator;
+
+  constructor(outputValidator: OutputValidator, txOut?: DeepPartial<Cardano.TxOut>) {
+    this.partialOutput = { ...txOut };
+    this.#outputValidator = outputValidator;
+  }
+
+  value(value: Cardano.Value): OutputBuilder {
+    this.partialOutput = { ...this.partialOutput, value };
+    return this;
+  }
+
+  coin(coin: Cardano.Lovelace): OutputBuilder {
+    this.partialOutput = { ...this.partialOutput, value: { ...this.partialOutput?.value, coins: coin } };
+    return this;
+  }
+
+  assets(assets: Cardano.TokenMap): OutputBuilder {
+    this.partialOutput = {
+      ...this.partialOutput,
+      value: { ...this.partialOutput?.value, assets }
+    };
+    return this;
+  }
+
+  /**
+   * @param assetId id
+   * @param quantity To remove an asset, set quantity to 0
+   */
+  asset(assetId: Cardano.AssetId, quantity: bigint): OutputBuilder {
+    const existingAssets = this.partialOutput?.value?.assets;
+    const assets: Cardano.TokenMap = isTokenMap(existingAssets)
+      ? existingAssets.set(assetId, quantity)
```

Comment:
Other methods do not mutate `partialOutput`. Let's keep it consistent and not mutate the `Map`. Instead, create a new map:
```suggestion
      ? new Map(existingAssets.entries()).set(assetId, quantity)
```